Realm/RLMProperty.mm

```diff
@@ -343,6 +343,9 @@ - (instancetype)initSwiftPropertyWithName:(NSString *)name
     }
     else if ([_objcRawType isEqualToString:@"@\"NSNumber\""]) {
         const char *numberType = [propertyValue objCType];
+        if (!numberType) {
+            @throw RLMException(@"Can't persist NSNumber directly: use a Swift-native number type.");
```

Comment:
Exactly. :+1:

>Is there any way we can provide proper error messages to Realm Swift users while maintaining backwards compatibility for Realm Objective-C users with Swift models?

No, I don't see how there would be a way as `RLMProperty` is agnostic about the class it belongs to. The only way would be making the initializer failable or catching the exception from the object schema creation and then rethrowing it, which I'd rather avoid just for that purpose.